allensdk/brain_observatory/ecephys/ecephys_project_api/http_engine.py

```diff
@@ -1,23 +1,211 @@
+
+    def stream(self, route):
+        """ Makes an http request and returns an iterator over the response.
+
+        Parameters
+        ----------
+        route :
+            the http route (under this object's host) to request against.
+
+        """
 
-    def stream(self, path):
-        url = f"{self.scheme}://{self.host}/{path}"
+        url = self._build_url(route)
         
+        start_time = time.perf_counter()
         response = requests.get(url, stream=True)
         response_mb = None
         if "Content-length" in response.headers:
             response_mb = float(response.headers["Content-length"]) / 1024 ** 2
 
-        # TODO: this should be async with write (response.raw.read_chunked?)
-        for ii, chunk in enumerate(response):
+        for ii, chunk in enumerate(response.iter_content(self.chunksize)):
             if ii == 0:
                 size_message = f"{response_mb:3.3}mb" if response_mb is not None else "potentially large"
                 logging.warning(f"downloading a {size_message} file from {url}")
-            yield chunk
\ No newline at end of file
+            yield chunk
+
+            elapsed = time.perf_counter() - start_time
+            if elapsed > self.timeout:
+                raise requests.Timeout(f"Download took {elapsed} seconds, but timeout was set to {self.timeout}")
+
+
+AsyncStreamCallbackType = Callable[[AsyncIterator[bytes]], Awaitable[None]]
+
+
+class AsyncHttpEngine(HttpEngine):
+
+    def __init__(
+        self, 
+        scheme: str, 
+        host: str, 
+        session: Optional[aiohttp.ClientSession] = None, 
+        **kwargs
+    ):
+        """ Simple tool for making asynchronous streaming http requests.
+
+        Parameters
+        ----------
+        scheme :
+            e.g "http" or "https"
+        host : 
+            will be used as the base for request urls
+        session : 
+            If provided, this preconstructed session will be used rather than 
+            a new one. Keep in mind that AsyncHttpEngine closes its session 
+            when it is garbage collected!
+        **kwargs :
+            Will be passed to parent.
+
+        """
+
+        super(AsyncHttpEngine, self).__init__(scheme, host, **kwargs)
+
+        if session:
+            self.session = session
+            warnings.warn(
+                "Recieved preconstructed session, ignoring timeout parameter."
+            )
+        else:
+            self.session = aiohttp.ClientSession(
+                timeout=aiohttp.client.ClientTimeout(self.timeout)
+            )
+
+    async def _stream_coroutine(
+        self, 
+        route: str, 
+        callback: AsyncStreamCallbackType
+    ):
+        url = self._build_url(route)
+
+        async with self.session.get(url) as response:
+            await callback(response.content.iter_chunked(self.chunksize))
+
+    def stream(
+        self, 
+        route: str
+    ) -> Callable[[AsyncStreamCallbackType], Awaitable[None]]:
```

Comment:
Would be helpful to note here that this stream can be consumed by `write_bytes_from_coroutine` to write the data to file.